build-plugins/build-support/src/main/kotlin/aws/sdk/kotlin/gradle/plugins/artifactmetrics/AnalyzeArtifactSizeMetricsTask.kt

```diff
@@ -0,0 +1,166 @@
+        metricsFile.convention(project.layout.buildDirectory.file(OUTPUT_PATH + "artifact-size-metrics.csv"))
+        analysisFile.convention(project.layout.buildDirectory.file(OUTPUT_PATH + "artifact-analysis.md"))
+        hasSignificantChangeFile.convention(project.layout.buildDirectory.file(OUTPUT_PATH + "has-significant-change.txt"))
+    }
+
+    @TaskAction
+    fun analyze() {
+        val latestReleaseMetricsFile =
+            project.layout.buildDirectory.file(OUTPUT_PATH + "latest-release-artifact-size-metrics.csv").get().asFile
+        writeLatestReleaseMetrics(latestReleaseMetricsFile)
+
+        val latestReleaseMetrics = latestReleaseMetricsFile.toMap()
+        val currentMetrics = metricsFile.get().asFile.toMap()
+        val analysis = analyzeArtifactSizeMetrics(latestReleaseMetrics, currentMetrics)
+
+        hasSignificantChangeFile.get().asFile.writeText(analysis.significantChange.toString())
+        val diffTable = createDiffTable(analysis)
+        val output = if (analysis.delta) diffTable else noDiffMessage
+
+        this.analysisFile.get().asFile.writeText(output)
+    }
+
+    private fun writeLatestReleaseMetrics(file: File) = runBlocking {
+        S3Client.fromEnvironment().use { s3 ->
+            s3.getObject(
+                GetObjectRequest {
+                    bucket = "" // TODO: Point to artifact size metrics bucket
+                    key = "" // TODO: Point to artifact size metrics for latest release
+                },
+            ) { latestReleaseMetrics ->
+                file.writeText(
+                    latestReleaseMetrics.body?.decodeToString() ?: throw GradleException("Metrics from latest release are empty"),
+                )
+            }
+        }
+    }
+
+    private fun analyzeArtifactSizeMetrics(
+        releaseMetrics: Map<String, Long>,
+        currentMetrics: Map<String, Long>,
+    ): ArtifactSizeMetricsAnalysis {
+        val pluginConfig = this.project.rootProject.extensions.getByType(ArtifactSizeMetricsPluginConfig::class.java)
+
+        val artifactNames = releaseMetrics.keys + currentMetrics.keys
+        val artifactSizeMetrics = artifactNames.associateWith { artifact ->
+            val current = currentMetrics[artifact] ?: 0
+            val release = releaseMetrics[artifact] ?: 0
+
+            val delta = current - release
+            val percentage = if (current == 0L || release == 0L) Double.NaN else delta.toDouble() / release.toDouble() * 100
+
+            ArtifactSizeMetric(
+                current,
+                release,
+                delta,
+                percentage,
+            )
+        }
+
+        val changeHappened = artifactSizeMetrics.values.any { it.delta != 0L }
+        val significantChange = artifactSizeMetrics.values.any {
+            (abs(it.percentage) > pluginConfig.significantChangeThresholdPercentage && it.delta > 0L) || // Increase in size above threshold
+                (it.latestReleaseSize == 0L) // New artifact
```

Comment:
question/discussion: Should we be blocking workflows on new artifacts? We don't have any baseline to compare them to. 

What should devs do when the workflow is blocked because a new artifact was added?

Reply:
I just realized Ian commented on this already, ignore this comment. https://github.com/awslabs/aws-kotlin-repo-tools/pull/34#discussion_r1548711066